--- closure-orig/src/com/google/javascript/jscomp/parsing/parser/TokenType.java	2015-02-27 11:55:33.525744051 -0600
+++ closure/src/com/google/javascript/jscomp/parsing/parser/TokenType.java	2015-02-27 12:02:32.257739345 -0600
@@ -144,6 +144,10 @@
   TEMPLATE_MIDDLE("template middle"),
   TEMPLATE_TAIL("template tail"),
   NO_SUBSTITUTION_TEMPLATE("no substitution template"),
+
+  // JAM begin
+  TRANSACTION,
+  // JAM end
   ;
 
   public final String value;

