--- closure-orig/src/com/google/javascript/jscomp/CrossModuleCodeMotion.java	2014-10-03 11:38:06.622298474 -0500
+++ closure/src/com/google/javascript/jscomp/CrossModuleCodeMotion.java	2014-10-02 17:21:03.097276387 -0500
@@ -233,6 +233,9 @@
         case Token.SWITCH:
         case Token.WHILE:
         case Token.FUNCTION:
+        // JAM begin
+        case Token.TRANSACTION:
+        // JAM end
           return true;
       }
     }

