--- closure-orig/src/com/google/javascript/jscomp/ControlFlowGraph.java	2015-02-27 11:55:33.521744051 -0600
+++ closure/src/com/google/javascript/jscomp/ControlFlowGraph.java	2015-02-27 11:55:37.809744003 -0600
@@ -162,6 +162,13 @@
         // control is going into the function's body, represented by the second
         // child.
         return n != parent.getFirstChild().getNext();
+      // JAM begin
+      case Token.TRANSACTION:
+        // When a transaction block is approached, the introspector
+        // expression is evaluated first (though the resulting value is
+        // not called), followed by the body.
+        return true;
+      // JAM end
       case Token.WHILE:
       case Token.DO:
       case Token.IF:

