--- closure-orig/src/com/google/javascript/jscomp/ControlFlowAnalysis.java	2015-02-27 11:55:33.525744051 -0600
+++ closure/src/com/google/javascript/jscomp/ControlFlowAnalysis.java	2015-02-27 11:55:37.809744003 -0600
@@ -254,6 +254,9 @@
         case Token.IF:
         case Token.WHILE:
         case Token.WITH:
+        // JAM begin
+        case Token.TRANSACTION:
+        // JAM end
           return n != parent.getFirstChild();
         case Token.DO:
           return n != parent.getFirstChild().getNext();
@@ -345,12 +348,28 @@
         return;
       case Token.LABEL:
         return;
+      // JAM begin
+      case Token.TRANSACTION:
+        handleTransaction(n);
+        return;
+      // JAM end
       default:
         handleStmt(n);
         return;
     }
   }
 
+  // JAM begin
+  private void handleTransaction(Node node) {
+    Node ispect = node.getFirstChild();
+    Node body = ispect.getNext();
+    createEdge(node, Branch.UNCOND, computeFallThrough(body));
+    connectToPossibleExceptionHandler(node, ispect);
+    // %%% Also need to model the potential calling of the handler, 
+    // %%% which is somewhat like exceptions.
+  }
+  // JAM end
+
   private void handleIf(Node node) {
     Node thenBlock = node.getFirstChild().getNext();
     Node elseBlock = thenBlock.getNext();

