--- closure-orig/src/com/google/javascript/jscomp/GlobalNamespace.java	2015-02-27 11:55:33.529744051 -0600
+++ closure/src/com/google/javascript/jscomp/GlobalNamespace.java	2015-02-27 11:55:37.809744003 -0600
@@ -651,6 +651,9 @@
           case Token.BITNOT:
           case Token.POS:
           case Token.NEG:
+          // JAM begin
+          case Token.TRANSACTION:
+          // JAM end
             break;
           case Token.CALL:
             if (n == parent.getFirstChild()) {
@@ -739,6 +742,9 @@
           case Token.BITNOT:
           case Token.POS:
           case Token.NEG:
+          // JAM begin
+          case Token.TRANSACTION:
+          // JAM end
             return Ref.Type.DIRECT_GET;
           case Token.HOOK:
             if (anc.getFirstChild() == prev) {

