--- closure-orig/src/com/google/javascript/jscomp/parsing/parser/Keywords.java	2015-02-27 11:55:33.525744051 -0600
+++ closure/src/com/google/javascript/jscomp/parsing/parser/Keywords.java	2015-02-27 12:02:32.253739345 -0600
@@ -74,6 +74,9 @@
     PUBLIC("public", TokenType.PUBLIC),
     STATIC("static", TokenType.STATIC),
     YIELD("yield", TokenType.YIELD),
+    // JAM begin
+    TRANSACTION("introspect", TokenType.TRANSACTION),
+    // JAM end
 
     //7.8 Literals
     NULL("null", TokenType.NULL),

