--- closure-orig/src/com/google/javascript/jscomp/TypeCheck.java	2014-10-03 11:38:06.662298473 -0500
+++ closure/src/com/google/javascript/jscomp/TypeCheck.java	2014-10-02 17:26:57.909267520 -0500
@@ -833,6 +833,9 @@
       case Token.DO:
       case Token.IF:
       case Token.WHILE:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         typeable = false;
         break;
 
