--- closure-orig/src/com/google/javascript/jscomp/CodeGenerator.java	2015-02-27 11:55:33.517744051 -0600
+++ closure/src/com/google/javascript/jscomp/CodeGenerator.java	2015-02-27 11:55:37.805744003 -0600
@@ -337,6 +337,22 @@
         }
         break;
 
+      // JAM begin
+      case Token.TRANSACTION:
+        add("introspect(");
+        add(first);
+        add(")");
+        if (context == Context.STATEMENT) {
+          addNonEmptyStatement(last, Context.OTHER, false);
+        } else {
+          cc.beginBlock();
+          add(last, Context.STATEMENT);
+          cc.endBlock();
+        }
+        Preconditions.checkState(childCount == 2);
+        break;
+      // JAM end
+
       case Token.FUNCTION:
         if (n.getClass() != Node.class) {
           throw new Error("Unexpected Node subclass.");

