--- closure-orig/src/com/google/javascript/jscomp/AstValidator.java	2014-10-03 11:38:06.614298474 -0500
+++ closure/src/com/google/javascript/jscomp/AstValidator.java	2014-10-02 17:12:47.653288768 -0500
@@ -162,6 +162,11 @@
       case Token.EXPORT:
         validateExport(n);
         return;
+      // JAM begin
+      case Token.TRANSACTION:
+        validateTransaction(n);
+        return;
+      // JAM end
       default:
         violation("Expected statement but was "
             + Token.name(n.getType()) + ".", n);
@@ -302,6 +307,12 @@
         validateYield(n);
         return;
 
+      // JAM begin
+      case Token.TRANSACTION:
+        validateTransaction(n);
+        return;
+      // JAM end
+
       default:
         violation("Expected expression but was "
             + Token.name(n.getType()), n);
@@ -827,6 +838,15 @@
     }
   }
 
+  // JAM begin
+  private void validateTransaction(Node n) {
+    validateNodeType(Token.TRANSACTION, n);
+    validateChildCount(n, 2);
+    validateExpression(n.getFirstChild());
+    validateBlock(n.getChildAtIndex(1));
+  }
+  // JAM end
+
   private void validateExprStmt(Node n) {
     validateNodeType(Token.EXPR_RESULT, n);
     validateChildCount(n, Token.arity(Token.EXPR_RESULT));

