--- closure-orig/src/com/google/javascript/jscomp/CrossModuleCodeMotion.java	2015-02-27 11:55:33.517744051 -0600
+++ closure/src/com/google/javascript/jscomp/CrossModuleCodeMotion.java	2015-02-27 11:55:37.809744003 -0600
@@ -232,6 +232,9 @@
         case Token.SWITCH:
         case Token.WHILE:
         case Token.FUNCTION:
+        // JAM begin
+        case Token.TRANSACTION:
+        // JAM end
           return true;
       }
     }

