--- closure-orig/src/com/google/javascript/jscomp/Compiler.java	2015-02-27 11:55:33.513744052 -0600
+++ closure/src/com/google/javascript/jscomp/Compiler.java	2015-02-27 11:55:37.805744003 -0600
@@ -2671,4 +2671,10 @@
   ImmutableMap<String, Node> getDefaultDefineValues() {
     return this.defaultDefineValues;
   }
+
+  // JAM begin
+  public static void close() {
+    compilerExecutor.shutdown();
+  }
+  // JAM end
 }

