--- closure-orig/src/com/google/javascript/jscomp/NodeUtil.java	2015-02-27 11:55:33.513744052 -0600
+++ closure/src/com/google/javascript/jscomp/NodeUtil.java	2015-02-27 12:04:41.537737892 -0600
@@ -970,6 +970,9 @@
       case Token.SWITCH:
       case Token.TRY:
       case Token.EMPTY:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         break;
 
       // Throws are by definition side effects, and yields are similar.
@@ -1422,6 +1425,9 @@
       case Token.SUPER:
       case Token.TRUE:
       case Token.TEMPLATELIT:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
       // Tokens from the type declaration AST
       case Token.STRING_TYPE:
       case Token.NUMBER_TYPE:
@@ -2002,6 +2008,9 @@
       case Token.SWITCH:
       case Token.CASE:
       case Token.DEFAULT_CASE:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         return true;
       default:
         return false;
@@ -2033,6 +2042,10 @@
         return parent.getFirstChild() != n;
       case Token.DEFAULT_CASE:
         return true;
+      // JAM begin
+      case Token.TRANSACTION:
+        return parent.getFirstChild() != n;
+      // JAM end
       default:
         Preconditions.checkState(isControlStructure(parent));
         return false;
@@ -3748,6 +3761,10 @@
         case Token.WHILE:
         case Token.DO:
           return false;
+        // JAM begin
+        case Token.TRANSACTION:
+          return parent.getFirstChild() == n;
+        // JAM end
         case Token.TRY:
           // Consider all code under a try/catch to be conditionally executed.
           if (!hasFinally(parent) || parent.getLastChild() != n) {

