--- closure-orig/src/com/google/javascript/jscomp/parsing/parser/trees/ParseTree.java	2015-02-27 11:55:33.529744051 -0600
+++ closure/src/com/google/javascript/jscomp/parsing/parser/trees/ParseTree.java	2015-02-27 12:02:32.257739345 -0600
@@ -125,6 +125,10 @@
     return (TemplateSubstitutionTree) this; }
   public ThisExpressionTree asThisExpression() { return (ThisExpressionTree) this; }
   public ThrowStatementTree asThrowStatement() { return (ThrowStatementTree) this; }
+  // JAM begin
+  public TransactionStatementTree asTransactionStatement() { return (TransactionStatementTree) this; }
+  public TransactionExpressionTree asTransactionExpression() { return (TransactionExpressionTree) this; }
+  // JAM end
   public TryStatementTree asTryStatement() { return (TryStatementTree) this; }
   public TypeNameTree asTypeName() { return (TypeNameTree) this; }
   public TypedParameterTree asTypedParameter() { return (TypedParameterTree) this; }

