--- closure-orig/src/com/google/javascript/jscomp/GlobalNamespace.java	2014-10-03 11:38:06.634298474 -0500
+++ closure/src/com/google/javascript/jscomp/GlobalNamespace.java	2014-10-02 17:22:20.957274441 -0500
@@ -649,6 +649,9 @@
           case Token.BITNOT:
           case Token.POS:
           case Token.NEG:
+          // JAM begin
+          case Token.TRANSACTION:
+          // JAM end
             break;
           case Token.CALL:
             if (n == parent.getFirstChild()) {
@@ -740,6 +743,9 @@
           case Token.BITNOT:
           case Token.POS:
           case Token.NEG:
+          // JAM begin
+          case Token.TRANSACTION:
+          // JAM end
             return Ref.Type.DIRECT_GET;
           case Token.HOOK:
             if (anc.getFirstChild() == prev) {
