--- closure-orig/src/com/google/javascript/jscomp/TypeCheck.java	2015-02-27 11:55:33.521744051 -0600
+++ closure/src/com/google/javascript/jscomp/TypeCheck.java	2015-02-27 12:02:32.261739345 -0600
@@ -804,6 +804,9 @@
       case Token.DO:
       case Token.IF:
       case Token.WHILE:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         typeable = false;
         break;
 

