--- closure-orig/src/com/google/javascript/jscomp/UnreachableCodeElimination.java	2015-02-27 11:55:33.525744051 -0600
+++ closure/src/com/google/javascript/jscomp/UnreachableCodeElimination.java	2015-02-27 12:02:32.261739345 -0600
@@ -209,6 +209,27 @@
         return;
       }
 
+      // JAM begin
+      // %%% This is not efficient.
+      // %%% Really need to update the unreachable code logic to mark
+      // %%% transaction introspector expressions as reachable.
+      Node np = n;
+      Node p = np.getParent(); 
+      while (p != null) {
+        if (p.getType() == Token.TRANSACTION) {
+          if (np == p.getFirstChild()) {
+            return;
+          } else {
+            // It should be fine to remove unreachable statements within
+            // the block.
+            break;
+          }
+        }
+        np = p;
+        p = np.getParent();
+      }
+      // JAM end
+
       switch (n.getType()) {
         // In the CFG, the only incoming edges the the DO node are from
         // breaks/continues and the condition. The edge from the previous

