--- closure-orig/src/com/google/javascript/jscomp/AstValidator.java	2015-02-27 11:55:33.521744051 -0600
+++ closure/src/com/google/javascript/jscomp/AstValidator.java	2015-02-27 11:55:37.805744003 -0600
@@ -162,6 +162,11 @@
       case Token.EXPORT:
         validateExport(n);
         return;
+      // JAM begin
+      case Token.TRANSACTION:
+        validateTransaction(n);
+        return;
+      // JAM end
       default:
         violation("Expected statement but was "
             + Token.name(n.getType()) + ".", n);
@@ -302,6 +307,12 @@
         validateYield(n);
         return;
 
+      // JAM begin
+      case Token.TRANSACTION:
+        validateTransaction(n);
+        return;
+      // JAM end
+
       default:
         violation("Expected expression but was "
             + Token.name(n.getType()), n);
@@ -826,6 +837,15 @@
     }
   }
 
+  // JAM begin
+  private void validateTransaction(Node n) {
+    validateNodeType(Token.TRANSACTION, n);
+    validateChildCount(n, 2);
+    validateExpression(n.getFirstChild());
+    validateBlock(n.getChildAtIndex(1));
+  }
+  // JAM end
+
   private void validateExprStmt(Node n) {
     validateNodeType(Token.EXPR_RESULT, n);
     validateChildCount(n);

