--- closure-orig/src/com/google/javascript/jscomp/ConstCheck.java	2015-02-27 11:55:33.517744051 -0600
+++ closure/src/com/google/javascript/jscomp/ConstCheck.java	2015-02-27 11:55:37.809744003 -0600
@@ -36,7 +36,9 @@
     implements CompilerPass {
 
   static final DiagnosticType CONST_REASSIGNED_VALUE_ERROR =
-      DiagnosticType.error(
+      // JAM begin
+      DiagnosticType.warning(
+      // JAM end
           "JSC_CONSTANT_REASSIGNED_VALUE_ERROR",
           "constant {0} assigned a value more than once.\n" +
           "Original definition at {1}");

