--- closure-orig/src/com/google/javascript/jscomp/VarCheck.java	2015-02-27 11:55:33.521744051 -0600
+++ closure/src/com/google/javascript/jscomp/VarCheck.java	2015-02-27 12:02:32.261739345 -0600
@@ -40,7 +40,9 @@
       "JSC_UNDEFINED_VARIABLE",
       "variable {0} is undeclared");
 
-  static final DiagnosticType VIOLATED_MODULE_DEP_ERROR = DiagnosticType.error(
+  // JAM: changed this to a warning rather than error since it was
+  // (seemingly incorrectly) triggered by a test case (cnn.com).
+  static final DiagnosticType VIOLATED_MODULE_DEP_ERROR = DiagnosticType.warning(
       "JSC_VIOLATED_MODULE_DEPENDENCY",
       "module {0} cannot reference {2}, defined in " +
       "module {1}, since {1} loads after {0}");

