--- closure-orig/lib/rhino/src/org/mozilla/javascript/Token.java	2013-08-19 15:26:58.276335368 -0500
+++ closure/lib/rhino/src/org/mozilla/javascript/Token.java	2014-02-04 13:44:47.339160392 -0600
@@ -226,7 +226,10 @@
         DEBUGGER       = 160,
         COMMENT        = 161,
         GENEXPR        = 162,
-        LAST_TOKEN     = 163;
+        // JAM begin
+        TRANSACTION    = 163,
+        // JAM end
+        LAST_TOKEN     = 164;
 
     /**
      * Returns a name for the token.  If Rhino is compiled with certain
@@ -411,6 +414,9 @@
           case DEBUGGER:        return "DEBUGGER";
           case COMMENT:         return "COMMENT";
           case GENEXPR:         return "GENEXPR";
+          // JAM begin
+          case TRANSACTION:     return "TRANSACTION";
+          // JAM end
         }
 
         // Token without name
@@ -457,6 +463,9 @@
             case Token.INSTANCEOF: return "instanceof";
             case Token.THROW:      return "throw";
             case Token.TRY:        return "try";
+            // JAM begin
+            case Token.TRANSACTION:return "introspect";
+            // JAM end
             default:               return null;
         }
     }
