--- closure-orig/src/com/google/javascript/jscomp/UnreachableCodeElimination.java	2013-08-19 15:26:58.320335366 -0500
+++ closure/src/com/google/javascript/jscomp/UnreachableCodeElimination.java	2013-08-30 01:45:24.186675747 -0500
@@ -209,6 +209,21 @@
         return;
       }
 
+      // JAM begin
+      // %%% This is not efficient.
+      // %%% Really need to update the unreachable code logic to mark
+      // %%% transaction handler expressions as reachable.
+      Node p = n; 
+      while (p != null) {
+        // %%% And it's part of the handler expression. It should be
+        // %%% fine to remove unreachable statements within the block.
+        if (p.getType() == Token.TRANSACTION) {
+          return;
+        }
+        p = p.getParent();
+      }
+      // JAM end
+
       switch (n.getType()) {
         // Removing an unreachable DO node is messy b/c it means we still have
         // to execute one iteration. If the DO's body has breaks in the middle,
