--- closure-orig/src/com/google/javascript/jscomp/Compiler.java	2014-10-23 11:43:01.851685920 -0500
+++ closure/src/com/google/javascript/jscomp/Compiler.java	2014-10-02 15:52:29.441409177 -0500
@@ -2742,4 +2742,10 @@
   ImmutableMap<String, Node> getDefaultDefineValues() {
     return this.defaultDefineValues;
   }
+
+  // JAM begin
+  public static void close() {
+    compilerExecutor.shutdown();
+  }
+  // JAM end
 }
