--- closure-orig/src/com/google/javascript/jscomp/VarCheck.java	2013-08-19 15:26:58.320335366 -0500
+++ closure/src/com/google/javascript/jscomp/VarCheck.java	2014-09-30 08:52:07.665360563 -0500
@@ -39,7 +39,9 @@
       "JSC_UNDEFINED_VARIABLE",
       "variable {0} is undeclared");
 
-  static final DiagnosticType VIOLATED_MODULE_DEP_ERROR = DiagnosticType.error(
+  // JAM: changed this to a warning rather than error since it was
+  // (seemingly incorrectly) triggered by a test case (cnn.com).
+  static final DiagnosticType VIOLATED_MODULE_DEP_ERROR = DiagnosticType.warning(
       "JSC_VIOLATED_MODULE_DEPENDENCY",
       "module {0} cannot reference {2}, defined in " +
       "module {1}, since {1} loads after {0}");
