--- closure-orig/src/com/google/javascript/jscomp/NodeUtil.java	2014-10-03 11:38:06.646298474 -0500
+++ closure/src/com/google/javascript/jscomp/NodeUtil.java	2014-10-02 17:26:11.565268678 -0500
@@ -931,6 +931,9 @@
       case Token.SWITCH:
       case Token.TRY:
       case Token.EMPTY:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         break;
 
       // Throws are by definition side effects
@@ -1981,6 +1984,9 @@
       case Token.SWITCH:
       case Token.CASE:
       case Token.DEFAULT_CASE:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         return true;
       default:
         return false;
@@ -2012,6 +2018,10 @@
         return parent.getFirstChild() != n;
       case Token.DEFAULT_CASE:
         return true;
+      // JAM begin
+      case Token.TRANSACTION:
+        return parent.getFirstChild() != n;
+      // JAM end
       default:
         Preconditions.checkState(isControlStructure(parent));
         return false;
@@ -3686,6 +3696,10 @@
         case Token.WHILE:
         case Token.DO:
           return false;
+        // JAM begin
+        case Token.TRANSACTION:
+          return parent.getFirstChild() == n;
+        // JAM end
         case Token.TRY:
           // Consider all code under a try/catch to be conditionally executed.
           if (!hasFinally(parent) || parent.getLastChild() != n) {
