--- closure-orig/src/com/google/javascript/jscomp/NodeUtil.java	2013-08-19 15:26:58.308335367 -0500
+++ closure/src/com/google/javascript/jscomp/NodeUtil.java	2013-09-28 18:46:58.710447165 -0500
@@ -812,6 +812,9 @@
       case Token.SWITCH:
       case Token.TRY:
       case Token.EMPTY:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         break;
 
       // Throws are by definition side effects
@@ -1706,6 +1709,9 @@
       case Token.SWITCH:
       case Token.CASE:
       case Token.DEFAULT_CASE:
+      // JAM begin
+      case Token.TRANSACTION:
+      // JAM end
         return true;
       default:
         return false;
@@ -1736,6 +1742,10 @@
         return parent.getFirstChild() != n;
       case Token.DEFAULT_CASE:
         return true;
+      // JAM begin
+      case Token.TRANSACTION:
+        return parent.getFirstChild() != n;
+      // JAM end
       default:
         Preconditions.checkState(isControlStructure(parent));
         return false;
@@ -3259,6 +3269,9 @@
           continue inspect;
         case Token.WHILE:
         case Token.DO:
+        // JAM begin
+        case Token.TRANSACTION:
+        // JAM end
           return false;
         case Token.TRY:
           // Consider all code under a try/catch to be conditionally executed.
