--- closure-orig/src/com/google/javascript/rhino/TokenStream.java	2013-08-19 15:26:58.328335366 -0500
+++ closure/src/com/google/javascript/rhino/TokenStream.java	2014-01-22 21:46:41.336397142 -0600
@@ -162,6 +162,9 @@
                 case 'o': X="boolean";id=true; break partial;
                 case 'r': X="private";id=true; break partial;
                 case 'x': X="extends";id=true; break partial;
+                // JAM begin
+                case 'u': X="suspend";id=true; break partial;
+                // JAM end
                 } break partial;
             case 8: switch (s.charAt(0)) {
                 case 'a': X="abstract";id=true; break partial;
@@ -175,10 +178,17 @@
                 else if (c=='p') { X="protected";id=true; }
                 else if (c=='t') { X="transient";id=true; }
                 break partial;
-            case 10: c=s.charAt(1);
-                if (c=='m') { X="implements";id=true; }
-                else if (c=='n') { X="instanceof";id=true; }
+            //case 10: c=s.charAt(1);
+            //    if (c=='m') { X="implements";id=true; }
+            //    else if (c=='n') { X="instanceof";id=true; }
+            //    break partial;
+            // JAM begin
+            case 10: c=s.charAt(2);
+                if (c=='p') { X="implements";id=true; }
+                else if (c=='s') { X="instanceof";id=true; }
+                else if (c=='t') { X="introspect";id=true; }
                 break partial;
+            // JAM end
             case 12: X="synchronized";id=true; break partial;
             }
             // partial match validate the entire string the one possibility
@@ -190,6 +200,11 @@
     public static boolean isJSIdentifier(String s) {
       int length = s.length();
 
+      // JAM begin
+      // %%% Could be more efficient if this is permanent.
+      if (length > 0 && (s.charAt(0) == '#' || s.charAt(0) == '`')) {
+      } else
+      // JAM end
       if (length == 0 ||
           Character.isIdentifierIgnorable(s.charAt(0)) ||
           !Character.isJavaIdentifierStart(s.charAt(0))) {
@@ -197,6 +212,11 @@
       }
 
       for (int i = 1; i < length; i++) {
+        // JAM begin
+        // %%% Maybe could be more efficient if this is permanent.
+        if (s.charAt(i) == '#') {
+        } else
+        // JAM end
         if (Character.isIdentifierIgnorable(s.charAt(i)) ||
             !Character.isJavaIdentifierPart(s.charAt(i))) {
           return false;
