--- closure-orig/src/com/google/javascript/jscomp/CodeGenerator.java	2014-10-03 11:38:06.618298474 -0500
+++ closure/src/com/google/javascript/jscomp/CodeGenerator.java	2014-10-02 17:14:38.897285988 -0500
@@ -334,6 +334,16 @@
         }
         break;
 
+      // JAM begin
+      case Token.TRANSACTION:
+        add("introspect(");
+        add(first);
+        add(")");
+        addNonEmptyStatement(last, Context.OTHER, false);
+        Preconditions.checkState(childCount == 2);
+        break;
+      // JAM end
+
       case Token.FUNCTION:
         if (n.getClass() != Node.class) {
           throw new Error("Unexpected Node subclass.");
